### ISSUE TITLE
NotarizeAsset (MsgServer): is not able to block a transaction from being consensed.

### ISSUE THREAD

**Issue by eckelj** [planetmint/planetmint-go#63]
CID asset creation messages are consensed but might not be valid for the planetmint application.
The message below is retrieved from the consensus; the data is just not stored in the key-value store. 

that's ok, but don't avoid spamming the network of actors that are not attested.

```
func (k msgServer) NotarizeAsset(goCtx context.Context, msg *types.MsgNotarizeAsset) (*types.MsgNotarizeAssetResponse, error) {
	ctx := sdk.UnwrapSDKContext(goCtx)

	_, found := k.machineKeeper.GetMachineIndex(ctx, msg.PubKey)

	if !found {
		return nil, errors.New("machine not found")
	}

	valid := ValidateSignature(msg.Hash, msg.Signature, msg.PubKey)
	if !valid {
		return nil, errors.New("invalid signature")
	}

	var asset = types.Asset{
		Hash:      msg.Hash,
		Signature: msg.Signature,
		Pubkey:    msg.PubKey,
	}

	k.StoreAsset(ctx, asset)

	return &types.MsgNotarizeAssetResponse{}, nil
}
```

**Comment by LaurentMontBlanc** [comment]
I've looked into this issue, especially in regards to the lifecycle of a TX in a Cosmos SDK application. The short answer is that the `msg server` is not nor will be able to block a transaction from being consensed.

The reason for that is that the only way of blocking an invalid `Tx` (or invalid `Msg` of a `Tx` in this case) from being consensed is through `CheckTx` and `DeliverTx`. These messages are called whenever a full-node receives a `Tx` and used to run *stateless* and *stateful* checks before adding `Txs` to the mempool. This means inclusion to a nodes mempool can only be blocked by these methods, they can still be found invalid at a later stage. 

The documentation has a pretty good summary of the Tx execution / state change (https://docs.cosmos.network/main/basics/tx-lifecycle#state-changes)

TL:DR; The msg server is called after the tx has already been added to the mempool (i.e.`DeliverTx`).

NOTE: Gas is only consumed for `DeliverTx`

Depending on what we prefer I see two options:
1. Allow for `DeliverTx` of invalid Tx Msgs and consume Gas for redistribution
2. Implement an AnteHandlerDecorator that is called during `CheckTx` but consume no gas for the checks

**Comment by LaurentMontBlanc** [comment]
According to the documentation transaction verification is just about the cryptographics and fees. Meaning only verifying that msg digest and signatures are correct as well as enough gas being provided for addition to the mempool.

Msg validation itself MUST be done by the msg server. A modules message server and keeper are solely responsible for updating the application state. In short this means it is valid for a transaction with the correct amount of gas and valid signatures to be added to a full-nodes mempool. 

In earlier implementations of Cosmos SDK application `ValidateBasic` was called during `CheckTx` to discard transactions holding invalid messages. However this approach is deprecated and discouraged.

I'm postponing this issue and figure out what kind of fees we want to charge for machines notarizing assets. I will then test how a custom `AnteHandler` affects application performance as machines are currently the only actors that are allowed to notarize.

If the `AnteHandler` approach does not suffice for our use case I will implement the gas consumption on msg validation.
